docs/devices.ipynb

```diff
@@ -0,0 +1,410 @@
+    "    cirq.PhasedXPowGate(phase_exponent=0.3)(q2),\n",
+    "    cirq_google.SYC(q1, q2),\n",
+    "    cirq_google.SYC(q2, q3),\n",
+    ")\n",
+    "my_device.validate_circuit(my_circuit)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "-sNjRTCKDb2D"
+   },
+   "source": [
+    "`my_circuit` satisfies all the device constraints and could be run on a Sycamore device."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "BrxB0eCUDvlI"
+   },
+   "source": [
+    "## 2. Metadata features\n",
+    "\n",
+    "Some devices will also expose additional information via the `metadata` property. You can look at the metadata information of the Sycamore device with:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "id": "clsvdcXzEOIV"
+   },
+   "outputs": [],
+   "source": [
+    "metadata = my_device.metadata\n",
+    "metadata"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "glMmeLevElt-"
+   },
+   "source": [
+    "The Sycamore device is a 2d grid device that exposes a `cirq.GridDeviceMetadata` with a uniform set of gates across all the qubits as well as a connectivity graph. You can explore the properties below:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "id": "9Zm_1IsFEiHT"
+   },
+   "outputs": [],
+   "source": [
+    "type(metadata)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "id": "JbwrPD_AETiX"
+   },
+   "outputs": [],
+   "source": [
+    "qubit_set = metadata.qubit_set\n",
+    "nx_graph = metadata.nx_graph\n",
+    "gateset = metadata.gateset\n",
+    "\n",
+    "print(gateset)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "fo6ZTYN4HBVL"
+   },
+   "source": [
+    "These metadata features can be useful when designing/building algorithms around certain device information in order to tailor them for that device."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "RvC4xz_OHT3s"
+   },
+   "source": [
+    "## 3. The `cirq.Device` interface\n",
+    "\n",
+    "For advanced users (such as vendors) it is also possible to implement your own Device with its own unique constraints and metadata information. Below you can implement our own fictituous device:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "id": "cA7AdVWJCpRr"
+   },
+   "outputs": [],
+   "source": [
+    "class MyDevice(cirq.Device):\n",
+    "  \"\"\"Five qubits on a line, supporting X/Y/Z and CZ between neighbors.\"\"\"\n",
+    "  def __init__(self):\n",
+    "      self._qubits = set(cirq.LineQubit.range(5))\n",
+    "      self._supported_gates = cirq.Gateset(\n",
+    "          cirq.XPowGate, cirq.YPowGate, cirq.ZPowGate, cirq.CZPowGate\n",
+    "      )\n",
+    "\n",
+    "  def validate_operation(self, operation):\n",
```

Comment:
Since this is supposed to be an instructive example, I would put good docstrings in here, such as "this validate_operation does blah, but you can customize to support..." and really make use of the docstrings to  help explain how to construct your own device and what the best practices are.  (For instance, in validate_circuit, you can explain that it is only needed if their are circuit-level constraints.)

Reply:
Added docstrings explaining behaviors in more detail